src/klr/klr.cpp

```diff
@@ -7,7 +7,143 @@ void GetModuleDirectory(HMODULE module, LPWSTR szPath)
 {
     DWORD dirLength = GetModuleFileName(module, szPath, MAX_PATH);
     for (dirLength--; dirLength >= 0 && szPath[dirLength] != '\\'; dirLength--);
-    szPath[dirLength + 1] = '\0';
+    szPath[dirLength + 1] = L'\0';
+}
+
+int LastIndexOfChar(LPCWSTR const pszStr, WCHAR c)
+{
+    int nIndex = lstrlen(pszStr) - 1;
+    do
+    {
+        if (pszStr[nIndex] == c)
+        {
+            break;
+        }
+    } while (--nIndex >= 0);
+
+    return nIndex;
+}
+
+bool StringsEqual(LPCWSTR const pszStrA, LPCWSTR const pszStrB)
+{
+    return ::_wcsicmp(pszStrA, pszStrB) == 0;
+}
+
+bool StringEndsWith(LPCWSTR const pszStr, LPCWSTR const pszSuffix)
+{
+    int nStrLen = lstrlen(pszStr);
+    int nSuffixLen = lstrlen(pszSuffix);
+    int nOffset = nStrLen - nSuffixLen;
+
+    if (nOffset < 0)
+    {
+        return false;
+    }
+
+    return StringsEqual(pszStr + nOffset, pszSuffix);
+}
+
+int LastPathSeparatorIndex(LPCWSTR const pszPath)
+{
+    int nLastSlashIndex = LastIndexOfChar(pszPath, L'/');
+    int nLastBackSlashIndex = LastIndexOfChar(pszPath, L'\\');
+    return max(nLastSlashIndex, nLastBackSlashIndex);
+}
+
+void GetParentDir(LPCWSTR const pszPath, LPWSTR const pszParentDir)
+{
+    int nLastSeparatorIndex = LastPathSeparatorIndex(pszPath);
+    if (nLastSeparatorIndex < 0)
+    {
+        StringCchCopyW(pszParentDir, MAX_PATH, L".");
```

Comment:
strcpy_s() works for char arrays but we need the version for wide char (WCHAR) arrays. The secure version of lstrcpy() is `StringCchCopyW()`:
http://msdn.microsoft.com/en-us/library/windows/desktop/ms647490(v=vs.85).aspx
http://msdn.microsoft.com/en-us/library/windows/desktop/ms647527(v=vs.85).aspx